### PR TITLE
Implemented timestamp for accrue interest

### LINKED ISSUE #174: use timestamp to accure interest

### PR DESCRIPTION
**Pull request by ayushmishra2005**
resolves #174

### CODE REVIEW

**Comment by yz89** [comment]
Get the delta time between the current block and last block firstly. Then calculate the interest rate in the range of blocks.

**Comment by yz89** [comment]
Record the timestamp of the previous block in storage.

File: pallets/loans/src/rate.rs [hunk 1]
```diff
@@ -41,13 +42,22 @@ impl From<Rate> for APR {
...
             .checked_div(&Rate::saturating_from_integer(block_per_year))
     }
 
-    pub fn accrued_interest_per_block(&self, amount: u128, block_per_year: u128) -> Option<u128> {
-        self.0.checked_mul_int(amount)?.checked_div(block_per_year)
+    pub fn accrued_interest_per_block(&self, amount: u128, delta_time: Timestamp) -> Option<u128> {
+        let fraction: Rate = Self::from_fraction(delta_time)?;
+        self.0
+            .checked_mul_int(amount)?
+            .checked_div(fraction.into_inner())
```
Comment (yz89) [inline pallets/loans/src/rate.rs:60]: What does this mean? Could you talk more about the functionality of this method?

**Comment by yz89** [comment]
If the unit test fails, the PR will not be merged.

File: pallets/loans/src/rate.rs (hunk 1 above)
Comment (ayushmishra2005) [reply pallets/loans/src/rate.rs]: Just corrected it. As discussed with you, it should be

```
   pub fn accrued_interest_per_block(&self, amount: u128, delta_time: Timestamp) -> Option<u128> {
        let fraction: Rate = Self::from_fraction(delta_time)?;
        self.0
            .checked_mul_int(amount)?
            .checked_mul(fraction.into_inner())
    }
```

Regarding `fraction.into_inner()`, I am converting FixedU128 into u128.

**Comment by ayushmishra2005** [comment]
> If the unit test fails, the PR will not be merged.

Yes @yz89 Please verify main code changes, after that I would work on test-cases.

File: pallets/loans/src/rate.rs [hunk 2]
```diff
@@ -41,13 +42,21 @@ impl From<Rate> for APR {
...
         self.0
             .checked_div(&Rate::saturating_from_integer(block_per_year))
     }
 
-    pub fn accrued_interest_per_block(&self, amount: u128, block_per_year: u128) -> Option<u128> {
-        self.0.checked_mul_int(amount)?.checked_div(block_per_year)
+    pub fn accrued_interest_per_block(&self, amount: u128, delta_time: Timestamp) -> Option<u128> {
+        let fraction: Rate = Self::from_fraction(delta_time)?;
+
+        fraction.checked_mul_int(self.0.checked_mul_int(amount)?)
```
Comment (ayushmishra2005) [inline pallets/loans/src/rate.rs:59]: @yz89 is this ok now?

Comment (yz89) [reply pallets/loans/src/rate.rs:59]: `let fraction: Rate = Rate::checked_from_rational(delta_time, SECONDS_PER_YEAR);`

Get fraction directly.

**Comment by yz89** [comment]
When fixing the unit test, you may need to set a reasonable timestamp for the block. https://github.com/paritytech/substrate/blob/master/frame/timestamp/src/lib.rs#L378

File: pallets/loans/src/rate.rs (hunk 2 above)
Comment (yz89) [reply pallets/loans/src/rate.rs:59]: > @yz89 is this ok now?

yes

File: pallets/loans/src/rate.rs [hunk 3]
```diff
@@ -41,13 +42,20 @@ impl From<Rate> for APR {
...
+    }
+
     pub fn rate_per_block(&self, block_per_year: u128) -> Option<Rate> {
         self.0
             .checked_div(&Rate::saturating_from_integer(block_per_year))
     }
 
-    pub fn accrued_interest_per_block(&self, amount: u128, block_per_year: u128) -> Option<u128> {
-        self.0.checked_mul_int(amount)?.checked_div(block_per_year)
+    pub fn accrued_interest_per_block(&self, amount: u128, delta_time: Timestamp) -> Option<u128> {
```
Comment (ayushmishra2005) [inline pallets/loans/src/rate.rs:56]: rename to accrued_interest

File: pallets/loans/src/rate.rs [hunk 4]
```diff
@@ -41,13 +42,20 @@ impl From<Rate> for APR {
 impl APR {
     pub const MAX: Rate = Rate::from_inner(350_000_000_000_000_000); // 35%
 
+    pub fn from_fraction(delta_time: Timestamp) -> Option<Rate> {
```
Comment (ayushmishra2005) [inline pallets/loans/src/rate.rs:45]: This is not needed. remove this.

**Review by yz89 — APPROVED** [review]

**Pull request merged**